### ISSUE TITLE
Leaderboard

### ISSUE THREAD

**Issue by weipipione** [weihaox/BrainHub#5]
Sorry to interrupt，I want to konw what brainencoder do you use in Leaderboard, brainx-v-1-4' or 'brainx'? And what is the difference between brainx-v-1-4' and 'brainx'？

**Comment by weihaox** [comment]
Hi @weipipione The current results are from `brainx-v-1-4`. The [captioning](https://github.com/weihaox/BrainHub/tree/main/caption/comparison/umbrae-v-1-4) and [grounding](https://github.com/weihaox/BrainHub/tree/main/bbox/umbrae-v-1-4) results for both checkpoints can be found in this repo. 

The difference lies in the sampling strategies. Please refer to Appendix D.3 *Sampling Strategies in Cross-Subject Training* in our [paper](https://arxiv.org/pdf/2404.07202) for more details.

**Comment by weipipione** [comment]
Thank you for your quick and useful response!

**Comment by weipipione** [comment]
Sorry to bother you again, Appendix D.3 Sampling Strategies in Cross-Subject Training describe four Sampling Strategies,including ‘Random’, ‘Stratified’, ‘Ours-R’ and ‘Ours’, but the code in train_brainx.py seems using θ=0.50, B=256 between tow random subjects, it a bit like 'Random', but why only two subjects in a Batch B instead of four is a bit confusing to me here. Can you open the code for all sampling Strategies?

**Comment by weihaox** [comment]
Hi @weipipione This Sampling Strategies part is added after rebuttal. The provided code as you see is the original implementation.  You can modify the [code](https://github.com/weihaox/UMBRAE/blob/main/umbrae/train_brainx.py#L245-L273) to load data from all four subjects into a single batch for training. The changes required are straightforward.

**Comment by weipipione** [comment]
Thank you for answering my confusion. I think I understand it.

**Comment by weipipione** [comment]
I am so sorry to bother you again, i try to modify the code to load data from all four subjects into a single batch for training, but the result about grounding is not satisfied. The modified code is as followed, I guess it made some mistakes, but I don't know how to correct them, can you help me check it? Thank you very much!

for epoch in progress_bar:
    voxel2emb.train()

    loss_recon_sum = 0.
    val_loss_recon_sum = 0.

    selected_sub = random.sample(subs, len(subs))
    selected_idx = [subs.index(item) for item in selected_sub]

    train_dl_0 = train_dls[selected_idx[0]]
    train_dl_1 = train_dls[selected_idx[1]]    
    train_dl_2 = train_dls[selected_idx[2]]
    train_dl_3 = train_dls[selected_idx[3]]

    for train_i, ((voxel_0, image_0), (voxel_1, image_1), (voxel_2, image_2), (voxel_3, image_3)) in enumerate(zip(train_dl_0, train_dl_1, train_dl_2, train_dl_3)):
        with torch.cuda.amp.autocast():
            optimizer.zero_grad()
            # voxel_0:[128, 3, 15724]; image_0:[128, 3, 256, 256] 
            repeat_index = train_i % 3
            voxel_0 = voxel_0[:,repeat_index].float()
            voxel_1 = voxel_1[:,repeat_index].float()
            voxel_2 = voxel_2[:,repeat_index].float()
            voxel_3 = voxel_3[:,repeat_index].float()

            nums = []
            if args.sampling_strategy=='Random':
                # 随机生成 3 个数
                nums = [random.randint(1, args.batch_size*2-1) for _ in range(3)]
                while sum(nums) >= args.batch_size*2 :
                    nums = [random.randint(1, args.batch_size*2-1) for _ in range(3)]

                # 计算第四个数
                nums.append(args.batch_size*2 - sum(nums))
                # 确保所有数都在 1 到 args.batch_size*2 之间，并且总和为 args.batch_size*2
                if all(1 <= num <= (args.batch_size*2-1) for num in nums) and sum(nums) == args.batch_size*2:
                    pass
                else:
                    # 如果不满足条件，抛出异常并中断程序
                    raise ValueError(f"生成的数不符合要求: {nums}. 总和必须为 args.batch_size*2")
                
                sample_indices_0 = torch.randint(0, voxel_0.size(0), (nums[0],))
                sample_indices_1 = torch.randint(0, voxel_1.size(0), (nums[1],))
                sample_indices_2 = torch.randint(0, voxel_2.size(0), (nums[2],))
                sample_indices_3 = torch.randint(0, voxel_3.size(0), (nums[3],))

            if args.sampling_strategy=='Stratified':
                sample_indices_0 = torch.randint(0, voxel_0.size(0), (args.batch_size//2,))
                sample_indices_1 = torch.randint(0, voxel_1.size(0), (args.batch_size//2,))
                sample_indices_2 = torch.randint(0, voxel_2.size(0), (args.batch_size//2,))
                sample_indices_3 = torch.randint(0, voxel_3.size(0), (args.batch_size//2,))

            if args.sampling_strategy=='Umbare_Random':
                # 随机生成 2 个数
                nums = [random.randint(1, args.batch_size) for _ in range(2)]
                while sum(nums) >= args.batch_size:
                    nums = [random.randint(1, args.batch_size) for _ in range(2)]

                nums.append(args.batch_size - sum(nums))
                # 确保所有数都在 1 到 args.batch_size 之间，并且总和为 args.batch_size
                if all(1 <= num <= (args.batch_size -1) for num in nums) and sum(nums) == args.batch_size:
                    pass
                else:
                    # 如果不满足条件，抛出异常并中断程序
                    raise ValueError(f"生成的数不符合要求: {nums}. 总和必须为 args.batch_size")
                 
                sample_indices_0 = torch.randint(0, voxel_0.size(0), (args.batch_size,))
                sample_indices_1 = torch.randint(0, voxel_1.size(0), (nums[0],))
                sample_indices_2 = torch.randint(0, voxel_2.size(0), (nums[1],))
                sample_indices_3 = torch.randint(0, voxel_3.size(0), (nums[2],))

            if args.sampling_strategy=='Umbare_Stratified':
                sample_indices_0 = torch.randint(0, voxel_0.size(0), (args.batch_size,))
                sample_indices_1 = torch.randint(0, voxel_1.size(0), (args.batch_size//3,))
                sample_indices_2 = torch.randint(0, voxel_2.size(0), (args.batch_size//3,))
                sample_indices_3 = torch.randint(0, voxel_3.size(0), (args.batch_size//3,))

            # # Select the samples from voxel_0,1,2,3
            voxel_0 = voxel_0[sample_indices_0]
            voxel_1 = voxel_1[sample_indices_1]
            voxel_2 = voxel_2[sample_indices_2]
            voxel_3 = voxel_3[sample_indices_3]

            # 从 image 中随机选取 nums 个样本
            image_0 = image_0[sample_indices_0]
            image_1 = image_1[sample_indices_1]
            image_2 = image_2[sample_indices_2]
            image_3 = image_3[sample_indices_3]

            #[128, 256, 1024]
            emb_voxel_0 = voxel2emb(voxel_0, modal=f'fmri{selected_sub[0]}') 
            emb_voxel_1 = voxel2emb(voxel_1, modal=f'fmri{selected_sub[1]}')  
            emb_voxel_2 = voxel2emb(voxel_2, modal=f'fmri{selected_sub[2]}')  
            emb_voxel_3 = voxel2emb(voxel_3, modal=f'fmri{selected_sub[3]}')  

            if args.use_image_aug:
                image_0 = img_augment(image_0) #[128, 3, 256, 256] -> [128, 3, 224, 224] 
                image_1 = img_augment(image_1) #[128, 3, 256, 256] -> [128, 3, 224, 224]
                image_2 = img_augment(image_2) #[128, 3, 256, 256] -> [128, 3, 224, 224]
                image_3 = img_augment(image_3) #[128, 3, 256, 256] -> [128, 3, 224, 224]


            emb_image_0 = image2emb.encode_image(image_0, 'image') # [128, 256, 1024] 
            emb_image_1 = image2emb.encode_image(image_1, 'image') # [128, 256, 1024] 
            emb_image_2 = image2emb.encode_image(image_2, 'image') # [128, 256, 1024] 
            emb_image_3 = image2emb.encode_image(image_3, 'image') # [128, 256, 1024] 


            if args.feat_dim == 4096:
                emb_image_0 = mm_projector(emb_image_0.to(torch.float16))
                emb_image_1 = mm_projector(emb_image_1.to(torch.float16))
                emb_image_2 = mm_projector(emb_image_2.to(torch.float16))
                emb_image_3 = mm_projector(emb_image_3.to(torch.float16))


            loss_recon = loss_fn(emb_voxel_0, emb_image_0) + loss_fn(emb_voxel_1, emb_image_1) + loss_fn(emb_voxel_2, emb_image_2) + loss_fn(emb_voxel_3, emb_image_3) 

            loss_recon_sum += loss_recon.item()
            loss = loss_recon 
            utils.check_loss(loss)
            accelerator.backward(loss)
            optimizer.step()

            losses.append(loss.item())
            lrs.append(optimizer.param_groups[0]['lr'])

            if args.lr_scheduler_type is not None:
                lr_scheduler.step()
